File: app/src/main/java/io/github/daylight00/nativeshell/NativePty.java
package io.github.daylight00.nativeshell;

import java.io.IOException;

final class NativePty {
    static {
        System.loadLibrary("shellbridge");
    }

    private NativePty() {
    }

    static native long spawn(
            String shellPath,
            String cwd,
            String home,
            String temporaryDirectory,
            int rows,
            int columns) throws IOException;

    static native int read(long handle, byte[] destination, int offset, int length) throws IOException;

    static native int write(long handle, byte[] source, int offset, int length) throws IOException;

    static native void resize(
            long handle,
            int rows,
            int columns,
            int pixelWidth,
            int pixelHeight) throws IOException;

    static native void signalProcessGroup(long handle, int signalNumber) throws IOException;

    static native int waitFor(long handle) throws IOException;

    static native void destroy(long handle);
}

File: app/src/main/kotlin/io/github/daylight00/androidterminal/NativePty.kt
package io.github.daylight00.androidterminal

import java.io.IOException

internal object NativePty {
    init {
        System.loadLibrary("shellbridge")
    }

    @Throws(IOException::class)
    external fun spawn(
        shellPath: String,
        cwd: String,
        home: String,
        temporaryDirectory: String,
        sharedStorageDirectory: String,
        rows: Int,
        columns: Int,
    ): Long

    @Throws(IOException::class)
    external fun read(handle: Long, destination: ByteArray, offset: Int, length: Int): Int

    @Throws(IOException::class)
    external fun write(handle: Long, source: ByteArray, offset: Int, length: Int): Int

    @Throws(IOException::class)
    external fun resize(
        handle: Long,
        rows: Int,
        columns: Int,
        pixelWidth: Int,
        pixelHeight: Int,
    )

    @Throws(IOException::class)
    external fun signalProcessGroup(handle: Long, signalNumber: Int)

    @Throws(IOException::class)
    external fun waitFor(handle: Long): Int

    external fun destroy(handle: Long)
}
